### ISSUE TITLE
Package caching

### ISSUE THREAD

**Issue by cohoe** [CFCC/TurboStacks#1]
From https://github.com/CFCC/TurboPuppet/issues/35

Sonatype Nexus (Yum and Choco)

Local package cache

**Comment by idontusenumbers** [comment]
The frustrating part about nexus is you gotta set it up on every client =/ Is there a work around? I suppose puppet makes that bit easier.

**Comment by cohoe** [comment]
I don't know if we can combine it with the lancache technology. But yeah, I suspect it's gonna require local config. Puppet "should" make that do-able for all of our managed hosts.

**Comment by cohoe** [comment]
This doesn't actually get us anything since the installers for everything are scattered across the internet. Only the metadata package can be cached and that's so small it's meaningless.

**Comment by idontusenumbers** [comment]
Are you sure? nexus works for me for java packages (i think?); I thought it proxied everything; does it not proxy the data files?

**Comment by cohoe** [comment]
Yeah for other package types (like Maven or even RPM/DEBs) the entire contents of the package are hosted. Chocolatey (NuGet) works by wrapping up install scripts but pointing to 3rd party hosting services for the actual Windows MSIs or EXEs. For example, `GoogleChrome` has you grab the installer from dl.google.com but feeds it into an install script hosted in the Chocolatey package.

**Comment by cohoe** [comment]
Ok maybe this isn't as dead as I thought. While I cannot cache the source mirrors, I can cache the download location. And that location can be a network share. https://docs.chocolatey.org/en-us/guides/usage/change-cache A quick test shows this actually works!

A hidden (pun intended) bonus: Choco is intended to be run as administrator which is reasonable. File shares can be mounted only within the administrator context letting us secretly mount drive letters without exposing them to campers to mess around in.
